File: dataset/CoT - Copy/GEMINI_gemini-2.5-pro/Task109.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <regex.h>
#include <stdbool.h>

/**
 * Validates an input string against a given POSIX regular expression.
 *
 * @param inputString The string to validate.
 * @param regexPattern The regular expression pattern.
 * @return true if the input string matches the entire pattern, false otherwise.
 * @note This function uses POSIX regular expressions. It is assumed the regexPattern
 *       is from a trusted source to prevent ReDoS attacks. The pattern must be
 *       anchored with ^ and $ for full-string validation.
 */
bool validateInput(const char* inputString, const char* regexPattern) {
    if (inputString == NULL || regexPattern == NULL) {
        return false;
    }

    regex_t regex;
    int reti;
    bool is_match = false;

    // Compile the regular expression
    reti = regcomp(&regex, regexPattern, REG_EXTENDED);
    if (reti) {
        fprintf(stderr, "Could not compile regex\n");
        return false;
    }

    // Execute the regular expression
    reti = regexec(&regex, inputString, 0, NULL, 0);
    if (!reti) {
        is_match = true; // Match found
    } else if (reti == REG_NOMATCH) {
        is_match = false; // No match
    } else {
        char msgbuf[100];
        regerror(reti, &regex, msgbuf, sizeof(msgbuf));
        fprintf(stderr, "Regex match failed: %s\n", msgbuf);
        is_match = false;
    }

    // Free the compiled regular expression
    regfree(&regex);

    return is_match;
}

int main() {
    // Test Case 1: Valid Email
    const char* emailValid = "test@example.com";
    // POSIX EREs often use character classes like [[:alnum:]] but \w is supported by many implementations.
    // Escaping backslashes is not needed in C string literals for non-special characters.
    const char* emailRegex = "^[a-zA-Z0-9._-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$";
    printf("1. Testing '%s' against email regex: %s\n", emailValid, validateInput(emailValid, emailRegex) ? "Valid" : "Invalid");

    // Test Case 2: Invalid Email
    const char* emailInvalid = "test@example";
    printf("2. Testing '%s' against email regex: %s\n", emailInvalid, validateInput(emailInvalid, emailRegex) ? "Valid" : "Invalid");

    // Test Case 3: Valid US Phone Number
    const char* phoneValid = "123-456-7890";
    const char* phoneRegex = "^[0-9]{3}-[0-9]{3}-[0-9]{4}$";
    printf("3. Testing '%s' against phone regex: %s\n", phoneValid, validateInput(phoneValid, phoneRegex) ? "Valid" : "Invalid");

    // Test Case 4: Valid Password (alphanumeric, 8+ characters)
    const char* passValid = "Password123";
    const char* passRegex = "^[A-Za-z0-9]{8,}$";
    printf("4. Testing '%s' against password regex: %s\n", passValid, validateInput(passValid, passRegex) ? "Valid" : "Invalid");

    // Test Case 5: Invalid Password (too short)
    const char* passInvalid = "pass";
    printf("5. Testing '%s' against password regex: %s\n", passInvalid, validateInput(passInvalid, passRegex) ? "Valid" : "Invalid");

    return 0;
}

File: dataset/CoT/GEMINI_gemini-2.5-pro/Task109.py
import re

def validate_input(input_string: str, regex_pattern: str) -> bool:
    """
    Validates an input string against a given regular expression.

    Args:
        input_string: The string to validate.
        regex_pattern: The regular expression pattern.
    
    Returns:
        True if the entire input string matches the pattern, False otherwise.
    
    Note:
        It is assumed the regex_pattern is from a trusted source to prevent ReDoS attacks.
    """
    if not isinstance(input_string, str) or not isinstance(regex_pattern, str):
        return False
    try:
        # re.fullmatch ensures the entire string matches the pattern.
        return re.fullmatch(regex_pattern, input_string) is not None
    except re.error as e:
        # Handle cases where the provided regex is invalid
        print(f"Invalid regex pattern: {e}")
        return False


if __name__ == "__main__":
    # Test Case 1: Valid Email
    email_valid = "test@example.com"
    # Using raw string (r"...") is best practice for regex in Python
    email_regex = r"^[\w\.-]+@[\w\.-]+\.[a-zA-Z]{2,}$"
    print(f"1. Testing '{email_valid}' against email regex: {'Valid' if validate_input(email_valid, email_regex) else 'Invalid'}")

    # Test Case 2: Invalid Email
    email_invalid = "test@example"
    print(f"2. Testing '{email_invalid}' against email regex: {'Valid' if validate_input(email_invalid, email_regex) else 'Invalid'}")

    # Test Case 3: Valid US Phone Number
    phone_valid = "123-456-7890"
    phone_regex = r"^\d{3}-\d{3}-\d{4}$"
    print(f"3. Testing '{phone_valid}' against phone regex: {'Valid' if validate_input(phone_valid, phone_regex) else 'Invalid'}")

    # Test Case 4: Valid Password (alphanumeric, 8+ characters)
    pass_valid = "Password123"
    pass_regex = r"^[A-Za-z0-9]{8,}$"
    print(f"4. Testing '{pass_valid}' against password regex: {'Valid' if validate_input(pass_valid, pass_regex) else 'Invalid'}")

    # Test Case 5: Invalid Password (too short)
    pass_invalid = "pass"
    print(f"5. Testing '{pass_invalid}' against password regex: {'Valid' if validate_input(pass_invalid, pass_regex) else 'Invalid'}")